lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -1030,3 +1041,135 @@ unsafeIntCast
+--
+-- == Notes on implementation choices
+--
+-- We cannot rely on cardano-api here because `Cardano.TxBodyContent BuildTx`
+-- cannot be extracted from an existing `TxBody`.
+--
+-- To avoid the need for `ledger -> wallet` conversions, this function can only
+-- be used to *add* tx body content.
+updateTx
+    :: forall era. IsRecentEra era
+    => Cardano.Tx era
+    -> TxUpdate
+    -> Either ErrUpdateSealedTx (Cardano.Tx era)
+updateTx (Cardano.Tx body existingKeyWits) extraContent = do
+    -- NOTE: The script witnesses are carried along with the cardano-api
+    -- `anyEraBody`.
+    body' <- modifyTxBody extraContent body
+
+    if null existingKeyWits
+       then Right $ Cardano.Tx body' mempty
+       else Left $ ErrExistingKeyWitnesses $ length existingKeyWits
+  where
+    era = recentEra @era
+
+    modifyTxBody
+        :: TxUpdate
+        -> Cardano.TxBody era
+        -> Either ErrUpdateSealedTx (Cardano.TxBody era)
+    modifyTxBody ebc = \case
+        Cardano.ShelleyTxBody shelleyEra bod scripts scriptData aux val ->
+            Right $ Cardano.ShelleyTxBody shelleyEra
+                (modifyShelleyTxBody ebc era bod)
+                (scripts ++ (flip toLedgerScript era
+                    <$> extraInputScripts))
+                scriptData
+                aux
+                val
+        Byron.ByronTxBody _ -> case Cardano.shelleyBasedEra @era of {}
+
+    TxUpdate _ _ _ extraInputScripts _ = extraContent
+
+    toLedgerScript
+        :: CA.Script CA.KeyHash
+        -> RecentEra era
+        -> Core.Script (ShelleyLedgerEra era)
+    toLedgerScript walletScript = \case
+        RecentEraBabbage ->
+            Cardano.toShelleyScript $ Cardano.ScriptInEra
+            Cardano.SimpleScriptInBabbage
+            (Cardano.SimpleScript $ toCardanoSimpleScript walletScript)
+        RecentEraConway ->
+            Cardano.toShelleyScript $ Cardano.ScriptInEra
+            Cardano.SimpleScriptInConway
+            (Cardano.SimpleScript $ toCardanoSimpleScript walletScript)
+
+modifyShelleyTxBody
+    :: forall era. TxUpdate
+    -> RecentEra era
+    -> TxBody (ShelleyLedgerEra era)
+    -> TxBody (ShelleyLedgerEra era)
+modifyShelleyTxBody txUpdate era = withConstraints era $
+    let
+
+        TxUpdate extraInputs extraCollateral extraOutputs _ feeUpdate = txUpdate
+        extraOutputs' = StrictSeq.fromList $ map toLedgerTxOut extraOutputs
+        extraInputs' =
+            Set.fromList (W.toLedger . fst <$> extraInputs)
+        extraCollateral' =
+            Set.fromList $ W.toLedger <$> extraCollateral
+        modifyFee old =
+            case feeUpdate of
+                UseNewTxFee c -> W.toLedger c
+                UseOldTxFee -> old
+
+    in
+        over feeTxBodyL modifyFee
+        . over outputsTxBodyL
+            (<> extraOutputs')
+        . over inputsTxBodyL
+            (<> extraInputs')
+        . over collateralInputsTxBodyL
+            (<> extraCollateral')
+  where
```

Comment:
@Anviking

Instead of having two sets of definitions (both a `where` clause and a `let` clause), can we coalesce both sets of definitions into a single set, in a single `where` clause?

Result: we need less wrapping, and save several lines.

```suggestion
    over feeTxBodyL modifyFee
    . over outputsTxBodyL
        (<> extraOutputs')
    . over inputsTxBodyL
        (<> extraInputs')
    . over collateralInputsTxBodyL
        (<> extraCollateral')
  where
    TxUpdate extraInputs extraCollateral extraOutputs _ feeUpdate = txUpdate
    extraOutputs' = StrictSeq.fromList $ map toLedgerTxOut extraOutputs
    extraInputs' = Set.fromList (W.toLedger . fst <$> extraInputs)
    extraCollateral' = Set.fromList $ W.toLedger <$> extraCollateral
        
    modifyFee old = case feeUpdate of
        UseNewTxFee c -> W.toLedger c
        UseOldTxFee -> old
```